tests/pass-dep/libc/libc-epoll.rs

```diff
@@ -84,7 +85,7 @@ fn test_epoll_socketpair() {
     assert_eq!(res, 0);
 
     // Check result from epoll_wait.
-    let expected_event = u32::try_from(libc::EPOLLRDHUP | libc::EPOLLIN | libc::EPOLLOUT).unwrap();
+    let expected_event = u32::try_from(libc::EPOLLRDHUP | libc::EPOLLIN).unwrap();
```

Comment:
For some reason I do not understand, this changes behavior in this test. @tiif any idea what is happening here? Sadly the test does not explain *why* EPOLLOUT should be triggered here. If I look at SocketPair::get_epoll_ready_events I am not sure why EPOLLOUT would be triggered -- no new space became available in the write buffer, after all.

Reply:
If I run the tests against real libc, I get yet another result here: `libc::EPOLLRDHUP | libc::EPOLLHUP | libc::EPOLLIN | libc::EPOLLOUT`. I also get other differences between the test and real libc, e.g. on line 172.

Reply:
This is the best explanation I could find for this behaviour. This is generally true for all the test that I have written for libc epoll in real system. 

> To be perhaps clearer, epoll_wait() won't return an fd unless something changed on that socket, but if something did change, it returns all the flags representing the current state.

Source: https://stackoverflow.com/questions/59288735/epoll-wait-return-epollout-even-with-epollet-flag

So since this file description is available to be written at the point when ``close`` is triggered, epoll will also mark it with ``EPOLLOUT`` to represent the current state of the file description. 

This is also documented in https://hackmd.io/@tiif/SJatftrH0#Epoll-semantics. Most of the design in this docs is pretty outdated, but I can go through it again and update it if it helps to serve as an alternative documentation.

Reply:
> So since this file description is available to be written at the point when close is triggered, epoll will also mark it with EPOLLOUT to represent the current state of the file description.

That doesn't answer the question though.
When the peer FD is closed, how can the FD be considered writeable?

There's already this very odd special case:
```
// This is an edge case. Whenever epollrdhup is triggered, epollin will be added
// even though there is no data in the buffer.
epoll_ready_events.epollin = true;
```
This sounds like we should also set `epollout = true`, tog et the old behavior back, where it worked just by accident.

But that doesn't explain why we should set it to true.

Reply:
> When the peer FD is closed, how can the FD be considered writeable?

Now I think about it, I agree with this. it shouldn't be considered writable although this is what libc returns. This should be considered part of the edge case behaviour instead of "it is writable". 

This indeed works by accident, because we notify the peer of closed fd to ``check_and_update_readiness`` before the fd actually closed. 

The logic is in ``socketpair::close``, in particular this chunk of code: 
```rust
        // Notify peer fd that closed has happened.
        if let Some(peer_fd) = self.peer_fd.upgrade() {
            // When any of the event happened, we check and update the status of all supported events
            // types of peer fd.
            peer_fd.check_and_update_readiness(ecx)?;
        }
```

Reply:
My assumption was for it to unblock all threads waiting on writing or reading to/from this, and letting them retry and subsequently fail because the FD was already closed.

Reply:
So the idea is -- if your peer gets closed, then a read can now return EOF, and a write can return EPIPE, and that's the sense in which sense there was a change in read status and write status?

Yeah I can see that. I'll update the comments.